service/history/historyReplicator.go

```diff
@@ -348,6 +353,123 @@ func (r *historyReplicator) ApplyEvents(ctx context.Context, request *h.Replicat
 	}
 }
 
+func (r *historyReplicator) checkDuplicatedReplicationEvent(context *workflowExecutionContext, logger bark.Logger) error {
+	_, err := context.loadWorkflowExecution()
+	if err == nil {
+		// Workflow execution already exist, looks like a duplicate start event, it is safe to ignore it
+		logger.Debugf("Dropping stale replication task for reset event.")
+		r.metricsClient.IncCounter(metrics.ReplicateHistoryEventsScope, metrics.DuplicateReplicationEventsCounter)
+		return nil
+	}
+	if _, ok := err.(*shared.EntityNotExistsError); !ok {
+		// GetWorkflowExecution failed with some transient error. Return err so we can retry the task later
+		return err
+	}
+	return nil
+}
+
+func (r *historyReplicator) ApplyResetEvent(ctx context.Context, request *h.ReplicateEventsRequest, logger bark.Logger) (retError error) {
```

Comment:
the logic added can work, only for the simple case, for complex case, such as conflict resolution (which requires version comparison), as well as return retry error with hints for re-sending missing kafka message, the new logic will not work.

moreover, the application of events to start the reseted workflow, as well as termination of the current running workflow, should be handled inside the state builder. 

history replicator should only handle the start event (because this is special, require create workflow & conflict resolution), common non start workflow event conflict resolution, update workflow execution and force termination of current running workflow for conflict resolution.

Reply:
I see, I also realized that yesterday.